digiroad2-oracle/src/test/scala/fi/liikennevirasto/digiroad2/service/lane/LaneServiceSpec.scala

```diff
@@ -0,0 +1,214 @@
+        EnclosedTrafficArea, None, None)
+
+      val existingLanes = ServiceWithDao.fetchExistingMainLanesByRoadLinks(Seq(mockRoadLink), Seq())
+      existingLanes.length should be(2)
+
+    }
+  }
+
+  test("Fetch existing Lanes by linksId"){
+    runWithRollback {
+
+      val newLane11 = ServiceWithDao.create(Seq(NewIncomeLane(0, 0, 500, 745, false, false, lanePropertiesValues11)), Set(100L), 1, usernameTest)
+      newLane11.length should be(1)
+
+      val newLane21 = ServiceWithDao.create(Seq(NewIncomeLane(0, 0, 500, 745, false, false, lanePropertiesValues21)), Set(100L), 2, usernameTest)
+      newLane21.length should be(1)
+
+      val newLane22 = ServiceWithDao.create(Seq(NewIncomeLane(0, 0, 500, 745, false, false, lanePropertiesValues22)), Set(100L), 2, usernameTest)
+      newLane22.length should be(1)
+
+      val existingLanes = ServiceWithDao.fetchExistingLanesByLinksIdAndSideCode(100L, 1)
+      existingLanes.length should be(1)
+
+      val existingLanesOtherSide = ServiceWithDao.fetchExistingLanesByLinksIdAndSideCode(100L, 2)
+      existingLanesOtherSide.length should be(2)
+
+    }
+  }
+
+
+  test("Update Lane Information") {
+    runWithRollback {
+
+      val updateValues11 = Seq( LaneProperty("lane_code", Seq(LanePropertyValue(11))),
+                            LaneProperty("lane_continuity", Seq(LanePropertyValue("2"))),
+                            LaneProperty("lane_type", Seq(LanePropertyValue("5"))),
+                            LaneProperty("lane_information", Seq(LanePropertyValue("12")))
+                          )
+
+
+      val newLane11 = ServiceWithDao.create(Seq(NewIncomeLane(0, 0, 500, 745, false, false, lanePropertiesValues11)), Set(100L), 1, usernameTest)
+      newLane11.length should be(1)
+
+      val updatedLane = ServiceWithDao.update(Seq(NewIncomeLane(newLane11.head, 0, 500, 745, false, false, updateValues11)), Set(100L), 1, usernameTest)
+      updatedLane.length should be(1)
+
+      }
+  }
+
+  test("Should not be able to Update main Lane") {
+    runWithRollback {
+
+      val updateValues11 = Seq( LaneProperty("lane_code", Seq(LanePropertyValue(12))),
+                            LaneProperty("lane_continuity", Seq(LanePropertyValue("2"))),
+                            LaneProperty("lane_type", Seq(LanePropertyValue("5"))),
+                            LaneProperty("lane_information", Seq(LanePropertyValue("12")))
+                          )
+
+
+      val newLane11 = ServiceWithDao.create(Seq(NewIncomeLane(0, 0, 500, 745, false, false, lanePropertiesValues11)), Set(100L), 1, usernameTest)
+      newLane11.length should be(1)
+
+      val thrown = intercept[IllegalArgumentException] {
+        ServiceWithDao.update(Seq(NewIncomeLane(newLane11.head, 0, 500, 745, false, false, updateValues11)), Set(100L), 1, usernameTest)
+      }
+
+      thrown.getMessage should be("Cannot change the code of main lane!")
+
+    }
+  }
+
+}
```

Comment:
I think is missing unit tests like for example:
- Split lane asset;
- Join two split lanes case they have equal values
- Should expire lane from deleted road links through the actor;
- Create new assets or update when existing side code adjustment on road links;
- Delete a simple sub lane (with historic)
- Delete sub lane in the middle of others lanes
- Delete a simple sub lane (without historic)

Reply:
Done